lib/Target/Btor/BtorToBtorIRTranslation.cpp

```diff
@@ -0,0 +1,534 @@
+      if (cur->args[i] > 0 
+      && !isValidChild(reachedLines.at(cur->args[i]))) {
+        continue;
+      }
+
+      if (cache.find(cur->args[i]) == cache.end()) {
+        if (cur->args[i] < 0) {
+          // if original operation is cached, negate it
+          if (cache.find(cur->args[i] * -1) != cache.end()) {
+            createNegateLine(cur->args[i], cache.at(cur->args[i] * -1)); 
+          } else {
+            todo.push_back(reachedLines.at(cur->args[i] * -1));
+          }
+        } else {
+          todo.push_back(reachedLines.at(cur->args[i]));
+        }
+      }
+    }
+    if (todo.size() != oldsize) {
+      continue;
+    }
+    if (!isValidChild(cur) 
+    || cache.find(cur->id) != cache.end()) {
+      todo.pop_back();
+      continue;
+    }
+    res = createMLIR(cur, cur->args);
+    assert(res && res->getNumResults() < 2 && res->getResult(0));
+    cache[cur->id] = res->getResult(0);
+    todo.pop_back();
+  }
+}
+
+OwningOpRef<FuncOp> Deserialize::buildInitFunction() {
+  // collect the return types for our init function
+  std::vector<Type> returnTypes(states.size(), nullptr);
+  for (uint32_t i = 0; i < states.size(); ++i) {
+    returnTypes[i] = builder.getIntegerType(states.at(i)->sort.bitvec.width);
+    assert(returnTypes[i]);
+  }
+  ArrayRef<Type> outputs(returnTypes);
+
+  // create init function signature
+  OperationState state(unknownLoc, FuncOp::getOperationName());
+  FuncOp::build(builder, state, "init",
+                FunctionType::get(context, {}, outputs));
+  OwningOpRef<FuncOp> funcOp = cast<FuncOp>(Operation::create(state));
+
+  // create basic block and accompanying builder
+  Region &region = funcOp->getBody();
+  OpBuilder::InsertionGuard guard(builder);
+  auto *body = builder.createBlock(&region);
+  builder.setInsertionPointToStart(body);
+
+  // clear cache so that values are mapped to the right Basic Block
+  cache.clear();
+  for (auto it = inits.begin(); it != inits.end(); ++it) {
+    toOp(*it);
+  }
+
+  // close with a fitting returnOp
+  std::vector<Value> testResults(states.size(), nullptr);
+  std::map<uint32_t, Value> undefOpsBySort;
+  uint32_t j = 0; // counters over inits vector
+  for (uint32_t i = 0, sz = states.size(); i < sz; ++i) {
+    if (states.at(i)->init > 0) {
+      // get the result of init's second argument since
+      // that is what we assign our state to  
+      testResults[i] = cache.at(inits.at(j++)->args[1]);
+    } else {
+      auto sort = returnTypes.at(i).getIntOrFloatBitWidth();
+      if (undefOpsBySort.find(sort) == undefOpsBySort.end()) {
+          auto res = builder.create<btor::UndefOp>(unknownLoc,
+                                    returnTypes.at(i));
+          assert(res && res->getNumResults() == 1);
+          undefOpsBySort[sort] = res->getResult(0);
+      }
+      testResults[i] = undefOpsBySort.at(sort);
+    }
+    assert(testResults[i]);
+  }
+  ArrayRef<Value> results(testResults);
```

Comment:
I am not sure this is needed. `ArrayRef` is a an adapter type, you should be able to use `testResults` directly